src/rnnlm/rnnlm-lattice-rescoring.cc

```diff
@@ -0,0 +1,204 @@
+//           2017 Hainan Xu
+//
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+
+#include "rnnlm/rnnlm-lattice-rescoring.h"
+#include "util/stl-utils.h"
+#include "util/text-utils.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+KaldiRnnlmDeterministicFst::~KaldiRnnlmDeterministicFst() {
+  int size = state_to_decodable_rnnlm_.size();
+  KALDI_ASSERT(state_to_nnet3_output_.size() == size);
+  for (int i = 0; i < size; i++) {
+    delete state_to_decodable_rnnlm_[i];
+    delete state_to_nnet3_output_[i];
+  }
+  
+  state_to_decodable_rnnlm_.resize(0);
+  state_to_nnet3_output_.resize(0);
+  state_to_wseq_.resize(0);
+  wseq_to_state_.clear();
+}
+
+void KaldiRnnlmDeterministicFst::Clear() {
+  int size = state_to_decodable_rnnlm_.size();
+  KALDI_ASSERT(state_to_nnet3_output_.size() == size);
+  for (int i = 1; i < size; i++) {
+    delete state_to_decodable_rnnlm_[i];
+    delete state_to_nnet3_output_[i];
+  }
+  
+  state_to_decodable_rnnlm_.resize(1);
+  state_to_nnet3_output_.resize(1);
+  state_to_wseq_.resize(1);
+  wseq_to_state_.clear();
+  wseq_to_state_[state_to_wseq_[0]] = 0;
+}
+
+void KaldiRnnlmDeterministicFst::ReadFstWordSymbolTableAndRnnWordlist(
+    const std::string &rnn_wordlist,
+    const std::string &word_symbol_table_rxfilename) {
+  // Reads symbol table.
+  fst::SymbolTable *fst_word_symbols = NULL;
+  if (!(fst_word_symbols =
+      fst::SymbolTable::ReadText(word_symbol_table_rxfilename))) {
+    KALDI_ERR << "Could not read symbol table from file "
+              << word_symbol_table_rxfilename;
+  }
+
+  full_voc_size_ = fst_word_symbols->NumSymbols();
+  fst_label_to_word_.resize(full_voc_size_);
+
+  for (int32 i = 0; i < fst_label_to_word_.size(); ++i) {
+    fst_label_to_word_[i] = fst_word_symbols->Find(i);
+    if (fst_label_to_word_[i] == "") {
+      KALDI_ERR << "Could not find word for integer " << i << "in the word "
+                << "symbol table, mismatched symbol table or you have discoutinuous "
+                << "integers in your symbol table?";
+    }
+  }
+
+  fst_label_to_rnn_label_.resize(fst_word_symbols->NumSymbols(), -1);
+
+  out_OOS_index_ = 1;
+  {
+    std::ifstream ifile(rnn_wordlist.c_str());
+    int32 id;
+    string word;
+    int32 i = 0;
+    while (ifile >> word >> id) {
+      if (word == "</s>") {
```

Comment:
IIRC, other parts of the code don't assume that these are the written forms of BOS and EOS; they take these as options.  The scripts do, currently, so this might not be fatal-- but generally speaking, programs don't take any arguments with written forms of words unless they absolutely have to for some reason, and this program breaks that pattern.  I'd be happier to add the same options as for the training tools, with --bos-symbol and --eos-symbol (and an ignored option, --brk-symbol, if it makes your life easier scripting-wise).